// @guidepup/guidepup/lib/macOS/KeyCodes.js
"use strict";
/* eslint-disable @typescript-eslint/no-duplicate-enum-values */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyCodes = void 0;
/**
 * macOS key codes.
 *
 * references:
 * - {@link https://eastmanreference.com/complete-list-of-applescript-key-codes}
 * - {@link http://macbiblioblog.blogspot.com/2014/12/key-codes-for-function-and-special-keys.html}
 */
var KeyCodes;
(function (KeyCodes) {
    /**
     * a
     */
    KeyCodes[KeyCodes["a"] = 0] = "a";
    KeyCodes[KeyCodes["A"] = 0] = "A";
    KeyCodes[KeyCodes["KeyA"] = 0] = "KeyA";
    /**
     * s
     */
    KeyCodes[KeyCodes["s"] = 1] = "s";
    KeyCodes[KeyCodes["S"] = 1] = "S";
    KeyCodes[KeyCodes["KeyS"] = 1] = "KeyS";
    /**
     * d
     */
    KeyCodes[KeyCodes["d"] = 2] = "d";
    KeyCodes[KeyCodes["D"] = 2] = "D";
    KeyCodes[KeyCodes["KeyD"] = 2] = "KeyD";
    /**
     * f
     */
    KeyCodes[KeyCodes["f"] = 3] = "f";
    KeyCodes[KeyCodes["F"] = 3] = "F";
    KeyCodes[KeyCodes["KeyF"] = 3] = "KeyF";
    /**
     * h
     */
    KeyCodes[KeyCodes["h"] = 4] = "h";
    KeyCodes[KeyCodes["H"] = 4] = "H";
    KeyCodes[KeyCodes["KeyH"] = 4] = "KeyH";
    /**
     * g
     */
    KeyCodes[KeyCodes["g"] = 5] = "g";
    KeyCodes[KeyCodes["G"] = 5] = "G";
    KeyCodes[KeyCodes["KeyG"] = 5] = "KeyG";
    /**
     * z
     */
    KeyCodes[KeyCodes["z"] = 6] = "z";
    KeyCodes[KeyCodes["Z"] = 6] = "Z";
    KeyCodes[KeyCodes["KeyZ"] = 6] = "KeyZ";
    /**
     * x
     */
    KeyCodes[KeyCodes["x"] = 7] = "x";
    KeyCodes[KeyCodes["X"] = 7] = "X";
    KeyCodes[KeyCodes["KeyX"] = 7] = "KeyX";
    /**
     * c
     */
    KeyCodes[KeyCodes["c"] = 8] = "c";
    KeyCodes[KeyCodes["C"] = 8] = "C";
    KeyCodes[KeyCodes["KeyC"] = 8] = "KeyC";
    /**
     * v
     */
    KeyCodes[KeyCodes["v"] = 9] = "v";
    KeyCodes[KeyCodes["V"] = 9] = "V";
    KeyCodes[KeyCodes["KeyV"] = 9] = "KeyV";
    /**
     * §
     */
    KeyCodes[KeyCodes["SectionSign"] = 10] = "SectionSign";
    /**
     * b
     */
    KeyCodes[KeyCodes["b"] = 11] = "b";
    KeyCodes[KeyCodes["B"] = 11] = "B";
    KeyCodes[KeyCodes["KeyB"] = 11] = "KeyB";
    /**
     * q
     */
    KeyCodes[KeyCodes["q"] = 12] = "q";
    KeyCodes[KeyCodes["Q"] = 12] = "Q";
    KeyCodes[KeyCodes["KeyQ"] = 12] = "KeyQ";
    /**
     * w
     */
    KeyCodes[KeyCodes["w"] = 13] = "w";
    KeyCodes[KeyCodes["W"] = 13] = "W";
    KeyCodes[KeyCodes["KeyW"] = 13] = "KeyW";
    /**
     * e
     */
    KeyCodes[KeyCodes["e"] = 14] = "e";
    KeyCodes[KeyCodes["E"] = 14] = "E";
    KeyCodes[KeyCodes["KeyE"] = 14] = "KeyE";
    /**
     * r
     */
    KeyCodes[KeyCodes["r"] = 15] = "r";
    KeyCodes[KeyCodes["R"] = 15] = "R";
    KeyCodes[KeyCodes["KeyR"] = 15] = "KeyR";
    /**
     * y
     */
    KeyCodes[KeyCodes["y"] = 16] = "y";
    KeyCodes[KeyCodes["Y"] = 16] = "Y";
    KeyCodes[KeyCodes["KeyY"] = 16] = "KeyY";
    /**
     * t
     */
    KeyCodes[KeyCodes["t"] = 17] = "t";
    KeyCodes[KeyCodes["T"] = 17] = "T";
    KeyCodes[KeyCodes["KeyT"] = 17] = "KeyT";
    /**
     * 1
     */
    KeyCodes[KeyCodes["Digit1"] = 18] = "Digit1";
    /**
     * 2
     */
    KeyCodes[KeyCodes["Digit2"] = 19] = "Digit2";
    /**
     * 3
     */
    KeyCodes[KeyCodes["Digit3"] = 20] = "Digit3";
    /**
     * 4
     */
    KeyCodes[KeyCodes["Digit4"] = 21] = "Digit4";
    /**
     * 6
     */
    KeyCodes[KeyCodes["Digit6"] = 22] = "Digit6";
    /**
     * 5
     */
    KeyCodes[KeyCodes["Digit5"] = 23] = "Digit5";
    /**
     * =
     */
    KeyCodes[KeyCodes["="] = 24] = "=";
    KeyCodes[KeyCodes["Equals"] = 24] = "Equals";
    /**
     * 9
     */
    KeyCodes[KeyCodes["Digit9"] = 25] = "Digit9";
    /**
     * 7
     */
    KeyCodes[KeyCodes["Digit7"] = 26] = "Digit7";
    /**
     * -
     */
    KeyCodes[KeyCodes["-"] = 27] = "-";
    KeyCodes[KeyCodes["Dash"] = 27] = "Dash";
    KeyCodes[KeyCodes["Minus"] = 27] = "Minus";
    /**
     * 8
     */
    KeyCodes[KeyCodes["Digit8"] = 28] = "Digit8";
    /**
     * 0
     */
    KeyCodes[KeyCodes["Digit0"] = 29] = "Digit0";
    /**
     * ]
     */
    KeyCodes[KeyCodes["]"] = 30] = "]";
    KeyCodes[KeyCodes["RightSquareBracket"] = 30] = "RightSquareBracket";
    /**
     * o
     */
    KeyCodes[KeyCodes["o"] = 31] = "o";
    KeyCodes[KeyCodes["O"] = 31] = "O";
    KeyCodes[KeyCodes["KeyO"] = 31] = "KeyO";
    /**
     * u
     */
    KeyCodes[KeyCodes["u"] = 32] = "u";
    KeyCodes[KeyCodes["U"] = 32] = "U";
    KeyCodes[KeyCodes["KeyU"] = 32] = "KeyU";
    /**
     * [
     */
    KeyCodes[KeyCodes["["] = 33] = "[";
    KeyCodes[KeyCodes["LeftSquareBracket"] = 33] = "LeftSquareBracket";
    /**
     * i
     */
    KeyCodes[KeyCodes["i"] = 34] = "i";
    KeyCodes[KeyCodes["I"] = 34] = "I";
    KeyCodes[KeyCodes["KeyI"] = 34] = "KeyI";
    /**
     * p
     */
    KeyCodes[KeyCodes["p"] = 35] = "p";
    KeyCodes[KeyCodes["P"] = 35] = "P";
    KeyCodes[KeyCodes["KeyP"] = 35] = "KeyP";
    /**
     * Return ↵
     *
     * This is actually the "Return" key for MacOS.
     */
    KeyCodes[KeyCodes["Enter"] = 36] = "Enter";
    KeyCodes[KeyCodes["Return"] = 36] = "Return";
    /**
     * l
     */
    KeyCodes[KeyCodes["l"] = 37] = "l";
    KeyCodes[KeyCodes["L"] = 37] = "L";
    KeyCodes[KeyCodes["KeyL"] = 37] = "KeyL";
    /**
     * j
     */
    KeyCodes[KeyCodes["j"] = 38] = "j";
    KeyCodes[KeyCodes["J"] = 38] = "J";
    KeyCodes[KeyCodes["KeyJ"] = 38] = "KeyJ";
    /**
     * '
     */
    KeyCodes[KeyCodes["'"] = 39] = "'";
    KeyCodes[KeyCodes["SingleQuote"] = 39] = "SingleQuote";
    /**
     * k
     */
    KeyCodes[KeyCodes["k"] = 40] = "k";
    KeyCodes[KeyCodes["K"] = 40] = "K";
    KeyCodes[KeyCodes["KeyK"] = 40] = "KeyK";
    /**
     * ;
     */
    KeyCodes[KeyCodes[";"] = 41] = ";";
    KeyCodes[KeyCodes["SemiColon"] = 41] = "SemiColon";
    /**
     * \
     */
    KeyCodes[KeyCodes["\\"] = 42] = "\\";
    KeyCodes[KeyCodes["Backslash"] = 42] = "Backslash";
    /**
     * ,
     */
    KeyCodes[KeyCodes[","] = 43] = ",";
    KeyCodes[KeyCodes["Comma"] = 43] = "Comma";
    /**
     * /
     */
    KeyCodes[KeyCodes["/"] = 44] = "/";
    KeyCodes[KeyCodes["ForwardSlash"] = 44] = "ForwardSlash";
    /**
     * n
     */
    KeyCodes[KeyCodes["n"] = 45] = "n";
    KeyCodes[KeyCodes["N"] = 45] = "N";
    KeyCodes[KeyCodes["KeyN"] = 45] = "KeyN";
    /**
     * m
     */
    KeyCodes[KeyCodes["m"] = 46] = "m";
    KeyCodes[KeyCodes["M"] = 46] = "M";
    KeyCodes[KeyCodes["KeyM"] = 46] = "KeyM";
    /**
     * .
     */
    KeyCodes[KeyCodes["."] = 47] = ".";
    KeyCodes[KeyCodes["Period"] = 47] = "Period";
    KeyCodes[KeyCodes["FullStop"] = 47] = "FullStop";
    /**
     * Tab ⇥
     */
    KeyCodes[KeyCodes["Tab"] = 48] = "Tab";
    /**
     * " "
     */
    KeyCodes[KeyCodes[" "] = 49] = " ";
    KeyCodes[KeyCodes["Space"] = 49] = "Space";
    KeyCodes[KeyCodes["Spacebar"] = 49] = "Spacebar";
    /**
     * `
     */
    KeyCodes[KeyCodes["`"] = 50] = "`";
    KeyCodes[KeyCodes["Backtick"] = 50] = "Backtick";
    KeyCodes[KeyCodes["Tilde"] = 50] = "Tilde";
    /**
     * Backspace ← / ⌫
     *
     * Labelled as Delete on Mac keyboards, this is actually
     * a Backspace key.
     */
    KeyCodes[KeyCodes["Backspace"] = 51] = "Backspace";
    // Delete = 51,
    /**
     * Line Feed ␊
     */
    KeyCodes[KeyCodes["LineFeed"] = 52] = "LineFeed";
    /**
     * Escape ⎋
     */
    KeyCodes[KeyCodes["Escape"] = 53] = "Escape";
    /**
     * Command ⌘
     */
    KeyCodes[KeyCodes["Command"] = 55] = "Command";
    KeyCodes[KeyCodes["CommandLeft"] = 55] = "CommandLeft";
    KeyCodes[KeyCodes["CommandRight"] = 55] = "CommandRight";
    KeyCodes[KeyCodes["Meta"] = 55] = "Meta";
    /**
     * ⇧
     */
    KeyCodes[KeyCodes["Shift"] = 56] = "Shift";
    KeyCodes[KeyCodes["ShiftLeft"] = 56] = "ShiftLeft";
    /**
     * Caps Lock ⇪
     */
    KeyCodes[KeyCodes["CapsLock"] = 57] = "CapsLock";
    /**
     * ⌥
     */
    KeyCodes[KeyCodes["Option"] = 58] = "Option";
    KeyCodes[KeyCodes["OptionLeft"] = 58] = "OptionLeft";
    KeyCodes[KeyCodes["Alt"] = 58] = "Alt";
    KeyCodes[KeyCodes["AltLeft"] = 58] = "AltLeft";
    /**
     * ⌃
     */
    KeyCodes[KeyCodes["Control"] = 59] = "Control";
    KeyCodes[KeyCodes["ControlLeft"] = 59] = "ControlLeft";
    /**
     * ⇧
     */
    KeyCodes[KeyCodes["ShiftRight"] = 60] = "ShiftRight";
    /**
     * ⌥
     */
    KeyCodes[KeyCodes["OptionRight"] = 61] = "OptionRight";
    KeyCodes[KeyCodes["AltRight"] = 61] = "AltRight";
    /**
     * ⌃
     */
    KeyCodes[KeyCodes["ControlRight"] = 62] = "ControlRight";
    /**
     * fn
     */
    KeyCodes[KeyCodes["Fn"] = 63] = "Fn";
    /**
     * F17
     */
    KeyCodes[KeyCodes["F17"] = 64] = "F17";
    /**
     * NumPad .
     */
    KeyCodes[KeyCodes["Decimal"] = 65] = "Decimal";
    /**
     * NumPad *
     */
    KeyCodes[KeyCodes["Multiply"] = 67] = "Multiply";
    /**
     * NumPad +
     */
    KeyCodes[KeyCodes["Add"] = 69] = "Add";
    KeyCodes[KeyCodes["Plus"] = 69] = "Plus";
    /**
     * NumPad Clear ⌧
     */
    KeyCodes[KeyCodes["Clear"] = 71] = "Clear";
    /**
     * Volume Up
     */
    KeyCodes[KeyCodes["VolumeUp"] = 72] = "VolumeUp";
    /**
     * Volume Down
     */
    KeyCodes[KeyCodes["VolumeDown"] = 73] = "VolumeDown";
    /**
     * Mute
     */
    KeyCodes[KeyCodes["Mute"] = 74] = "Mute";
    /**
     * NumPad /
     */
    KeyCodes[KeyCodes["Divide"] = 75] = "Divide";
    /**
     * ⌅
     *
     * Not used.
     */
    // Enter = 76,
    /**
     * NumPad -
     */
    KeyCodes[KeyCodes["Subtract"] = 78] = "Subtract";
    /**
     * F18
     */
    KeyCodes[KeyCodes["F18"] = 79] = "F18";
    /**
     * F19
     */
    KeyCodes[KeyCodes["F19"] = 80] = "F19";
    /**
     * NumPad =
     *
     * Not used.
     */
    // Equals = 81,
    /**
     * NumPad 0
     *
     * Not used.
     */
    // Digit0 = 82,
    /**
     * NumPad 1
     *
     * Not used.
     */
    // Digit1 = 83,
    /**
     * NumPad 2
     *
     * Not used.
     */
    // Digit2 = 84,
    /**
     * NumPad 3
     *
     * Not used.
     */
    // Digit3 = 85,
    /**
     * NumPad 4
     *
     * Not used.
     */
    // Digit4 = 86,
    /**
     * NumPad 5
     *
     * Not used.
     */
    // Digit5 = 87,
    /**
     * NumPad 6
     *
     * Not used.
     */
    // Digit6 = 88,
    /**
     * NumPad 7
     *
     * Not used.
     */
    // Digit7 = 89,
    /**
     * F20
     */
    KeyCodes[KeyCodes["F20"] = 90] = "F20";
    /**
     * NumPad 8
     *
     * Not used.
     */
    // Digit8 = 91,
    /**
     * NumPad 9
     *
     * Not used.
     */
    // Digit9 = 92,
    /**
     * F5
     */
    KeyCodes[KeyCodes["F5"] = 96] = "F5";
    /**
     * F6
     */
    KeyCodes[KeyCodes["F6"] = 97] = "F6";
    /**
     * F7
     */
    KeyCodes[KeyCodes["F7"] = 98] = "F7";
    /**
     * F3
     */
    KeyCodes[KeyCodes["F3"] = 99] = "F3";
    /**
     * F8
     */
    KeyCodes[KeyCodes["F8"] = 100] = "F8";
    /**
     * F9
     */
    KeyCodes[KeyCodes["F9"] = 101] = "F9";
    /**
     * F11
     */
    KeyCodes[KeyCodes["F11"] = 103] = "F11";
    /**
     * F13
     */
    KeyCodes[KeyCodes["F13"] = 105] = "F13";
    /**
     * F16
     */
    KeyCodes[KeyCodes["F16"] = 106] = "F16";
    /**
     * F14
     */
    KeyCodes[KeyCodes["F14"] = 107] = "F14";
    /**
     * F10
     */
    KeyCodes[KeyCodes["F10"] = 109] = "F10";
    /**
     * F12
     */
    KeyCodes[KeyCodes["F12"] = 111] = "F12";
    /**
     * F15
     */
    KeyCodes[KeyCodes["F15"] = 113] = "F15";
    /**
     * Help / Insert
     */
    KeyCodes[KeyCodes["Help"] = 114] = "Help";
    KeyCodes[KeyCodes["Insert"] = 114] = "Insert";
    /**
     * Home
     */
    KeyCodes[KeyCodes["Home"] = 115] = "Home";
    /**
     * Page Up
     */
    KeyCodes[KeyCodes["PageUp"] = 116] = "PageUp";
    /**
     * Forward Delete ⌦
     */
    KeyCodes[KeyCodes["ForwardDelete"] = 117] = "ForwardDelete";
    KeyCodes[KeyCodes["Delete"] = 117] = "Delete";
    /**
     * F4
     */
    KeyCodes[KeyCodes["F4"] = 118] = "F4";
    /**
     * end
     */
    KeyCodes[KeyCodes["End"] = 119] = "End";
    /**
     * F2
     */
    KeyCodes[KeyCodes["F2"] = 120] = "F2";
    /**
     * Page Down
     */
    KeyCodes[KeyCodes["PageDown"] = 121] = "PageDown";
    /**
     * F1
     */
    KeyCodes[KeyCodes["F1"] = 122] = "F1";
    /**
     * ←
     */
    KeyCodes[KeyCodes["Left"] = 123] = "Left";
    KeyCodes[KeyCodes["ArrowLeft"] = 123] = "ArrowLeft";
    /**
     * →
     */
    KeyCodes[KeyCodes["Right"] = 124] = "Right";
    KeyCodes[KeyCodes["ArrowRight"] = 124] = "ArrowRight";
    /**
     * ↓
     */
    KeyCodes[KeyCodes["Down"] = 125] = "Down";
    KeyCodes[KeyCodes["ArrowDown"] = 125] = "ArrowDown";
    /**
     * ↑
     */
    KeyCodes[KeyCodes["Up"] = 126] = "Up";
    KeyCodes[KeyCodes["ArrowUp"] = 126] = "ArrowUp";
    /**
     * F4
     *
     * Not used.
     */
    // F4 = 131,
    /**
     * F3
     *
     * Not used.
     */
    // F3 = 160,
})(KeyCodes || (exports.KeyCodes = KeyCodes = {}));
