// @guidepup/guidepup/lib/macOS/VoiceOver/keyCodeCommands.js
"use strict";
// REF: https://www.apple.com/voiceover/info/guide/_1131.html
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyCodeCommands = void 0;
const KeyCodes_1 = require("../KeyCodes");
const Modifiers_1 = require("../Modifiers");
const VO = [Modifiers_1.Modifiers.Control, Modifiers_1.Modifiers.Option];
/**
 * Object of key code commands for the VoiceOver screen reader on MacOS.
 *
 * Reference: {@link https://www.apple.com/voiceover/info/guide/_1131.html}
 */
exports.keyCodeCommands = {
    start: {
        keyCode: KeyCodes_1.KeyCodes.F5,
        modifiers: [Modifiers_1.Modifiers.Command],
        description: "Start VoiceOver",
        representation: "Command-F5",
    },
    quit: {
        keyCode: KeyCodes_1.KeyCodes.F5,
        modifiers: [Modifiers_1.Modifiers.Command],
        description: "Quit VoiceOver",
        representation: "Command-F5",
    },
    // General Commands
    toggleLock: {
        keyCode: KeyCodes_1.KeyCodes.SemiColon,
        modifiers: VO,
        description: "Lock and unlock the VO (Control and Option) keys",
        representation: "VO-;",
    },
    openUtilityMenu: {
        keyCode: KeyCodes_1.KeyCodes.F8,
        modifiers: VO,
        description: "Open VoiceOver Utility",
        representation: "VO-F8",
    },
    openHelpMenu: {
        keyCode: KeyCodes_1.KeyCodes.H,
        modifiers: VO,
        description: "Open the VoiceOver Help menu",
        representation: "VO-H",
    },
    openQuickStart: {
        keyCode: KeyCodes_1.KeyCodes.F8,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Open the VoiceOver Quick Start",
        representation: "VO-Command-F8",
    },
    openOnlineHelp: {
        keyCode: KeyCodes_1.KeyCodes.ForwardSlash,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Open VoiceOver online help",
        representation: "VO-Shift-/",
    },
    startKeyboardHelp: {
        keyCode: KeyCodes_1.KeyCodes.k,
        modifiers: VO,
        description: "Start keyboard help",
        representation: "VO-K",
    },
    hearItemDescription: {
        keyCode: KeyCodes_1.KeyCodes.n,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Hear a description of the item in the VoiceOver cursor",
        representation: "VO-Shift-N",
    },
    openCommandsMenu: {
        keyCode: [KeyCodes_1.KeyCodes.h, KeyCodes_1.KeyCodes.h],
        modifiers: VO,
        description: "Open the Commands menu",
        representation: "VO-H-H",
    },
    openFindMenu: {
        keyCode: KeyCodes_1.KeyCodes.f,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Open the Find menu",
        representation: "VO-Shift-F",
    },
    stopAction: {
        keyCode: KeyCodes_1.KeyCodes.Escape,
        description: "Close a menu or rotor, stop an action, or exit a mode",
        representation: "Escape",
    },
    ignoreNextKeyCombination: {
        keyCode: KeyCodes_1.KeyCodes.Tab,
        modifiers: VO,
        description: "Tell VoiceOver to ignore the next key combination you press",
        representation: "VO-Tab",
    },
    openVerbosityRotor: {
        keyCode: KeyCodes_1.KeyCodes.v,
        modifiers: VO,
        description: "Open the verbosity rotor",
        representation: "VO-V",
    },
    magnifyItem: {
        keyCode: KeyCodes_1.KeyCodes.RightSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Magnify the item in the VoiceOver cursor",
        representation: "VO-}",
    },
    shrinkItem: {
        keyCode: KeyCodes_1.KeyCodes.LeftSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Shrink the item in the VoiceOver cursor",
        representation: "VO-{",
    },
    toggleCursorAndPanels: {
        keyCode: KeyCodes_1.KeyCodes.F11,
        modifiers: VO,
        description: "Temporarily hide or show the VoiceOver cursor and the caption or braille panels",
        representation: "VO-F11",
    },
    toggleCaptionPanel: {
        keyCode: KeyCodes_1.KeyCodes.F10,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hide or show the caption panel only",
        representation: "VO-Command-F10",
    },
    resizeOrMoveCaptionPanel: {
        keyCode: KeyCodes_1.KeyCodes.F10,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Resize or move the caption panel",
        representation: "VO-Shift-F10",
    },
    toggleBraillePanel: {
        keyCode: KeyCodes_1.KeyCodes.F9,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hide or show the braille panel only",
        representation: "VO-Command-F9",
    },
    resizeOrMoveBraillePanel: {
        keyCode: KeyCodes_1.KeyCodes.F9,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Resize or move the braille panel",
        representation: "VO-Shift-F9",
    },
    tileVisuals: {
        keyCode: KeyCodes_1.KeyCodes.F10,
        modifiers: VO,
        description: "Tile visuals (dim the screen, highlight the caption or braille panel, and show the item in the VoiceOver cursor in the center of the screen).",
        representation: "VO-F10",
    },
    toggleKeyboardCommander: {
        keyCode: KeyCodes_1.KeyCodes.k,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Enable or disable the Keyboard Commander",
        representation: "VO-Shift-K",
    },
    toggleScreenCurtain: {
        keyCode: KeyCodes_1.KeyCodes.F11,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Turn the screen black (screen curtain)",
        representation: "VO-Shift-F11",
    },
    cycleRightThroughSpeechSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Cycle through speech settings (rate, pitch, volume, intonation, voice)",
        representation: "VO-Command-Shift-Right Arrow",
    },
    cycleLeftThroughSpeechSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Cycle through speech settings (rate, pitch, volume, intonation, voice)",
        representation: "VO-Command-Shift-Left Arrow",
    },
    changeUpCurrentSpeechSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Change the current speech setting (rate, pitch, volume, intonation, voice)",
        representation: "VO-Command-Shift-Up Arrow",
    },
    changeDownCurrentSpeechSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Change the current speech setting (rate, pitch, volume, intonation, voice)",
        representation: "VO-Command-Shift-Down Arrow",
    },
    pressAndReleaseMouse: {
        keyCode: KeyCodes_1.KeyCodes.Space,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Press and release mouse button",
        representation: "VO-Shift-Space bar",
    },
    // Interaction Commands
    interactWithItem: {
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Interact with an item",
        representation: "VO-Shift-Down Arrow",
    },
    stopInteractingWithItem: {
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Stop interacting with an item",
        representation: "VO-Shift-Up Arrow",
    },
    performDefaultActionForItem: {
        keyCode: KeyCodes_1.KeyCodes.Space,
        modifiers: VO,
        description: "Perform the default action for the item in the VoiceOver cursor",
        representation: "VO-Space bar",
    },
    selectItem: {
        keyCode: KeyCodes_1.KeyCodes.Return,
        modifiers: VO,
        description: "Select a menu or list item",
        representation: "VO-Return",
    },
    selectMultipleItems: {
        keyCode: KeyCodes_1.KeyCodes.Space,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Select multiple items",
        representation: "VO-Command-Space bar",
    },
    toggleStickyMouse: {
        keyCode: KeyCodes_1.KeyCodes.Space,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Perform a sticky mouse down or mouse up (for use when dragging an item from one location to drop in another location)",
        representation: "VO-Command-Shift-Space bar",
    },
    doubleClick: {
        keyCode: [KeyCodes_1.KeyCodes.Space, KeyCodes_1.KeyCodes.Space],
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Click the item under the mouse cursor",
        representation: "VO-Shift-Space bar-Space bar",
    },
    toggleDisclosureTriangle: {
        keyCode: KeyCodes_1.KeyCodes.Backslash,
        modifiers: VO,
        description: "Open or close a disclosure triangle",
        representation: "VO-\\",
    },
    readTableRow: {
        keyCode: KeyCodes_1.KeyCodes.r,
        modifiers: VO,
        description: "Read a row in a table",
        representation: "VO-R",
    },
    readTableColumn: {
        keyCode: [KeyCodes_1.KeyCodes.c, KeyCodes_1.KeyCodes.c],
        modifiers: VO,
        description: "Read a column in a table",
        representation: "VO-C-C",
    },
    readTableColumnHeader: {
        keyCode: KeyCodes_1.KeyCodes.c,
        modifiers: VO,
        description: "Read the column header in a table",
        representation: "VO-C",
    },
    readTableRowAndColumnNumbers: {
        keyCode: KeyCodes_1.KeyCodes.t,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read row and column numbers in a table",
        representation: "VO-Shift-T",
    },
    sortTableColumn: {
        keyCode: KeyCodes_1.KeyCodes.Backslash,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Sort a column in a table",
        representation: "VO-|",
    },
    interactWithScrollbars: {
        keyCode: KeyCodes_1.KeyCodes.s,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Interact with scroll bars",
        representation: "VO-Shift-S",
    },
    resizeObject: {
        keyCode: KeyCodes_1.KeyCodes.Backtick,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Resize a window or an object",
        representation: "VO-~",
    },
    moveObject: {
        keyCode: KeyCodes_1.KeyCodes.Backtick,
        modifiers: VO,
        description: "Move a window or an object",
        representation: "VO-`",
    },
    // Navigation Commands
    moveUp: {
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: VO,
        description: "Move up",
        representation: "VO-Up Arrow",
    },
    moveDown: {
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: VO,
        description: "Move down",
        representation: "VO-Down Arrow",
    },
    moveToPrevious: {
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: VO,
        description: "Move to previous",
        representation: "VO-Left Arrow",
    },
    moveToNext: {
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: VO,
        description: "Move to next",
        representation: "VO-Right Arrow",
    },
    moveToVisibleAreaTop: {
        keyCode: KeyCodes_1.KeyCodes.Home,
        modifiers: VO,
        description: "Move to the top of the visible area (such as a window or text area) where the VoiceOver cursor is located",
        representation: "VO-Home",
    },
    moveToVisibleAreaBottom: {
        keyCode: KeyCodes_1.KeyCodes.End,
        modifiers: VO,
        description: "Move to the bottom of the visible area (such as a window or text area) where the VoiceOver cursor is located",
        representation: "VO-End",
    },
    moveToAreaTop: {
        keyCode: KeyCodes_1.KeyCodes.Home,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to the top of the area (such as a window or text area) where the VoiceOver cursor is located, scrolling if necessary",
        representation: "VO-Shift-Home",
    },
    moveToAreaBottom: {
        keyCode: KeyCodes_1.KeyCodes.End,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to the bottom of the area (such as a window or text area) where the VoiceOver cursor is located, scrolling if necessary",
        representation: "VO-Shift-End",
    },
    moveToFirst: {
        keyCode: KeyCodes_1.KeyCodes.Home,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the top of a window, the first item in the Dock, or the first item on your desktop, depending on your location",
        representation: "VO-Command-Home",
    },
    moveToLast: {
        keyCode: KeyCodes_1.KeyCodes.End,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the lower-right corner of a window, the last item in the Dock, or the last item on your desktop, depending on your location",
        representation: "VO-Command-End",
    },
    moveToFrontWindow: {
        keyCode: KeyCodes_1.KeyCodes.F2,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to the front the window where the VoiceOver cursor is located and make it active",
        representation: "VO-Shift-F2",
    },
    closeWindow: {
        keyCode: KeyCodes_1.KeyCodes.F2,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Close the window where the VoiceOver cursor is located",
        representation: "VO-Command-F2",
    },
    openItemChooser: {
        keyCode: KeyCodes_1.KeyCodes.i,
        modifiers: VO,
        description: "Open the Item Chooser",
        representation: "VO-I",
    },
    moveToDock: {
        keyCode: KeyCodes_1.KeyCodes.d,
        modifiers: VO,
        description: "Move to the desktop",
        representation: "VO-D",
    },
    moveToDesktop: {
        keyCode: KeyCodes_1.KeyCodes.d,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to the desktop",
        representation: "VO-D",
    },
    moveToMenuBar: {
        keyCode: KeyCodes_1.KeyCodes.m,
        modifiers: VO,
        description: "Move to the menu bar",
        representation: "VO-M",
    },
    moveToFirstStatusMenuInMenuBar: {
        keyCode: [KeyCodes_1.KeyCodes.m, KeyCodes_1.KeyCodes.m],
        modifiers: VO,
        description: "Move to the first status menu in the menu bar",
        representation: "VO-M-M",
    },
    openSpotlightMenu: {
        keyCode: [KeyCodes_1.KeyCodes.m, KeyCodes_1.KeyCodes.m, KeyCodes_1.KeyCodes.m],
        modifiers: VO,
        description: "Open the Spotlight menu",
        representation: "VO-M-M-M",
    },
    openShortcutMenu: {
        keyCode: KeyCodes_1.KeyCodes.m,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Open a shortcut menu",
        representation: "VO-Shift-J",
    },
    jumpToLinkedItem: {
        keyCode: KeyCodes_1.KeyCodes.j,
        modifiers: VO,
        description: "Jump to a linked item (for example, from a Mail message in the Inbox to its message text)",
        representation: "VO-J",
    },
    toggleCursorTrackingOptions: {
        keyCode: KeyCodes_1.KeyCodes.F3,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Temporarily disable or enable the cursor tracking options you selected in VoiceOver Utility. The command doesn't change the settings in VoiceOver Utility.",
        representation: "VO-Shift-F3",
    },
    moveCursorToKeyboardFocus: {
        keyCode: KeyCodes_1.KeyCodes.F4,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move VoiceOver cursor to keyboard focus",
        representation: "VO-Shift-F4",
    },
    moveKeyboardFocusToCursor: {
        keyCode: KeyCodes_1.KeyCodes.F4,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move keyboard focus to VoiceOver cursor",
        representation: "VO-Command-F4",
    },
    moveCursorToMouseFocus: {
        keyCode: KeyCodes_1.KeyCodes.F5,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move VoiceOver cursor to mouse cursor",
        representation: "VO-Shift-F5",
    },
    moveMouseFocusToCursor: {
        keyCode: KeyCodes_1.KeyCodes.F5,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move mouse cursor to VoiceOver cursor",
        representation: "VO-Command-F5",
    },
    jumpCommand: {
        keyCode: KeyCodes_1.KeyCodes.j,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Jump command",
        representation: "VO-Shift-J",
    },
    jumpToTopEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Jump to the top edge of an area. Used with jump command",
        representation: "VO-Shift-Up Arrow",
    },
    jumpToRightEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Jump to the right edge of an area. Used with jump command",
        representation: "VO-Shift-Right Arrow",
    },
    jumpToBottomEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Jump to the bottom edge of an area. Used with jump command",
        representation: "VO-Shift-Down Arrow",
    },
    jumpToLeftEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Jump to the left edge of an area. Used with jump command",
        representation: "VO-Shift-Left Arrow",
    },
    jumpToTopVisibleEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: VO,
        description: "Jump to the top visible edge of an area. Used with jump command",
        representation: "VO-Up Arrow",
    },
    jumpToRightVisibleEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: VO,
        description: "Jump to the right visible edge of an area. Used with jump command",
        representation: "VO-Right Arrow",
    },
    jumpToBottomVisibleEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: VO,
        description: "Jump to the bottom visible edge of an area. Used with jump command",
        representation: "VO-Down Arrow",
    },
    jumpToLeftVisibleEdge: {
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: VO,
        description: "Jump to the left visible edge of an area. Used with jump command",
        representation: "VO-Left Arrow",
    },
    jumpBeforeSplitter: {
        keyCode: KeyCodes_1.KeyCodes.LeftSquareBracket,
        modifiers: VO,
        description: "Jump to the area that precedes a horizontal or vertical splitter",
        representation: "VO-[",
    },
    jumpAfterSplitter: {
        keyCode: KeyCodes_1.KeyCodes.RightSquareBracket,
        modifiers: VO,
        description: "Jump to the area that follows a horizontal or vertical splitter",
        representation: "VO-]",
    },
    findText: {
        keyCode: KeyCodes_1.KeyCodes.f,
        modifiers: VO,
        description: "Find text",
        representation: "VO-F",
    },
    navigateUp: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Navigate in given direction, wrapping when necessary",
        representation: "VO-Command-Up Arrow",
    },
    cycleRightThroughNavigationSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Cycle through navigation settings (Headings, Form Controls, Landmarks, etc.)",
        representation: "VO-Command-Right Arrow",
    },
    navigateDown: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Navigate in given direction, wrapping when necessary",
        representation: "VO-Command-Down Arrow",
    },
    cycleLeftThroughNavigationSettings: {
        // VoiceOver documentation incorrect
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Cycle through navigation settings (Headings, Form Controls, Landmarks, etc.)",
        representation: "VO-Command-Left Arrow",
    },
    toggleHotSpot1: {
        keyCode: KeyCodes_1.KeyCodes.Digit1,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 1",
        representation: "VO-Shift-1",
    },
    toggleHotSpot2: {
        keyCode: KeyCodes_1.KeyCodes.Digit2,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 2",
        representation: "VO-Shift-2",
    },
    toggleHotSpot3: {
        keyCode: KeyCodes_1.KeyCodes.Digit3,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 3",
        representation: "VO-Shift-3",
    },
    toggleHotSpot4: {
        keyCode: KeyCodes_1.KeyCodes.Digit4,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 4",
        representation: "VO-Shift-4",
    },
    toggleHotSpot5: {
        keyCode: KeyCodes_1.KeyCodes.Digit5,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 5",
        representation: "VO-Shift-5",
    },
    toggleHotSpot6: {
        keyCode: KeyCodes_1.KeyCodes.Digit6,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 6",
        representation: "VO-Shift-6",
    },
    toggleHotSpot7: {
        keyCode: KeyCodes_1.KeyCodes.Digit7,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 7",
        representation: "VO-Shift-7",
    },
    toggleHotSpot8: {
        keyCode: KeyCodes_1.KeyCodes.Digit8,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 8",
        representation: "VO-Shift-8",
    },
    toggleHotSpot9: {
        keyCode: KeyCodes_1.KeyCodes.Digit9,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 9",
        representation: "VO-Shift-9",
    },
    toggleHotSpot0: {
        keyCode: KeyCodes_1.KeyCodes.Digit0,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Toggle hot spot 0",
        representation: "VO-Shift-0",
    },
    jumpToHotSpot1: {
        keyCode: KeyCodes_1.KeyCodes.Digit1,
        modifiers: VO,
        description: "Jump to hot spot 1",
        representation: "VO-1",
    },
    jumpToHotSpot2: {
        keyCode: KeyCodes_1.KeyCodes.Digit2,
        modifiers: VO,
        description: "Jump to hot spot 2",
        representation: "VO-2",
    },
    jumpToHotSpot3: {
        keyCode: KeyCodes_1.KeyCodes.Digit3,
        modifiers: VO,
        description: "Jump to hot spot 3",
        representation: "VO-3",
    },
    jumpToHotSpot4: {
        keyCode: KeyCodes_1.KeyCodes.Digit4,
        modifiers: VO,
        description: "Jump to hot spot 4",
        representation: "VO-4",
    },
    jumpToHotSpot5: {
        keyCode: KeyCodes_1.KeyCodes.Digit5,
        modifiers: VO,
        description: "Jump to hot spot 5",
        representation: "VO-5",
    },
    jumpToHotSpot6: {
        keyCode: KeyCodes_1.KeyCodes.Digit6,
        modifiers: VO,
        description: "Jump to hot spot 6",
        representation: "VO-6",
    },
    jumpToHotSpot7: {
        keyCode: KeyCodes_1.KeyCodes.Digit7,
        modifiers: VO,
        description: "Jump to hot spot 7",
        representation: "VO-7",
    },
    jumpToHotSpot8: {
        keyCode: KeyCodes_1.KeyCodes.Digit8,
        modifiers: VO,
        description: "Jump to hot spot 8",
        representation: "VO-8",
    },
    jumpToHotSpot9: {
        keyCode: KeyCodes_1.KeyCodes.Digit9,
        modifiers: VO,
        description: "Jump to hot spot 9",
        representation: "VO-9",
    },
    jumpToHotSpot0: {
        keyCode: KeyCodes_1.KeyCodes.Digit0,
        modifiers: VO,
        description: "Jump to hot spot 0",
        representation: "VO-0",
    },
    describeHotSpot1: {
        keyCode: KeyCodes_1.KeyCodes.Digit1,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 1",
        representation: "VO-Command-1",
    },
    describeHotSpot2: {
        keyCode: KeyCodes_1.KeyCodes.Digit2,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 2",
        representation: "VO-Command-2",
    },
    describeHotSpot3: {
        keyCode: KeyCodes_1.KeyCodes.Digit3,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 3",
        representation: "VO-Command-3",
    },
    describeHotSpot4: {
        keyCode: KeyCodes_1.KeyCodes.Digit4,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 4",
        representation: "VO-Command-4",
    },
    describeHotSpot5: {
        keyCode: KeyCodes_1.KeyCodes.Digit5,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 5",
        representation: "VO-Command-5",
    },
    describeHotSpot6: {
        keyCode: KeyCodes_1.KeyCodes.Digit6,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 6",
        representation: "VO-Command-6",
    },
    describeHotSpot7: {
        keyCode: KeyCodes_1.KeyCodes.Digit7,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 7",
        representation: "VO-Command-7",
    },
    describeHotSpot8: {
        keyCode: KeyCodes_1.KeyCodes.Digit8,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 8",
        representation: "VO-Command-8",
    },
    describeHotSpot9: {
        keyCode: KeyCodes_1.KeyCodes.Digit9,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 9",
        representation: "VO-Command-9",
    },
    describeHotSpot0: {
        keyCode: KeyCodes_1.KeyCodes.Digit0,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Hear a description of hot spot 0",
        representation: "VO-Command-0",
    },
    monitorHotSpot1: {
        keyCode: KeyCodes_1.KeyCodes.Digit1,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 1",
        representation: "VO-Command-Shift-1",
    },
    monitorHotSpot2: {
        keyCode: KeyCodes_1.KeyCodes.Digit2,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 2",
        representation: "VO-Command-Shift-2",
    },
    monitorHotSpot3: {
        keyCode: KeyCodes_1.KeyCodes.Digit3,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 3",
        representation: "VO-Command-Shift-3",
    },
    monitorHotSpot4: {
        keyCode: KeyCodes_1.KeyCodes.Digit4,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 4",
        representation: "VO-Command-Shift-4",
    },
    monitorHotSpot5: {
        keyCode: KeyCodes_1.KeyCodes.Digit5,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 5",
        representation: "VO-Command-Shift-5",
    },
    monitorHotSpot6: {
        keyCode: KeyCodes_1.KeyCodes.Digit6,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 6",
        representation: "VO-Command-Shift-6",
    },
    monitorHotSpot7: {
        keyCode: KeyCodes_1.KeyCodes.Digit7,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 7",
        representation: "VO-Command-Shift-7",
    },
    monitorHotSpot8: {
        keyCode: KeyCodes_1.KeyCodes.Digit8,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 8",
        representation: "VO-Command-Shift-8",
    },
    monitorHotSpot9: {
        keyCode: KeyCodes_1.KeyCodes.Digit9,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 9",
        representation: "VO-Command-Shift-9",
    },
    monitorHotSpot0: {
        keyCode: KeyCodes_1.KeyCodes.Digit0,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Monitor hot spot 0",
        representation: "VO-Command-Shift-0",
    },
    jumpToParentFolder: {
        keyCode: KeyCodes_1.KeyCodes.Backslash,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Jump back to a parent folder",
        representation: "VO-Command-\\",
    },
    // Orientation Commands
    hearApplicationSummary: {
        keyCode: KeyCodes_1.KeyCodes.F1,
        modifiers: VO,
        description: "Hear the application summary",
        representation: "VO-F1",
    },
    openApplicationChooser: {
        keyCode: [KeyCodes_1.KeyCodes.F1, KeyCodes_1.KeyCodes.F1],
        modifiers: VO,
        description: "Open the Application Chooser",
        representation: "VO-F1-F1",
    },
    hearWindowSummary: {
        keyCode: KeyCodes_1.KeyCodes.F2,
        modifiers: VO,
        description: "Hear the window summary",
        representation: "VO-F2",
    },
    openWindowChooser: {
        keyCode: [KeyCodes_1.KeyCodes.F2, KeyCodes_1.KeyCodes.F2],
        modifiers: VO,
        description: "Open the Window Chooser",
        representation: "VO-F2-F2",
    },
    describeItem: {
        keyCode: KeyCodes_1.KeyCodes.F3,
        modifiers: VO,
        description: "Describe the item in the VoiceOver cursor",
        representation: "VO-F3",
    },
    describeItemSize: {
        keyCode: KeyCodes_1.KeyCodes.F3,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Describe the size of the item in the VoiceOver cursor",
        representation: "VO-Command-F3",
    },
    describeItemPosition: {
        keyCode: [KeyCodes_1.KeyCodes.F3, KeyCodes_1.KeyCodes.F3],
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Describe the position of the item in the VoiceOver cursor",
        representation: "VO-Command-F3-F3",
    },
    describeItemWithKeyboardFocus: {
        keyCode: KeyCodes_1.KeyCodes.F4,
        modifiers: VO,
        description: "Describe the item that has the keyboard focus",
        representation: "VO-F4",
    },
    describeLocationOfInsertionPoint: {
        keyCode: [KeyCodes_1.KeyCodes.F4, KeyCodes_1.KeyCodes.F4],
        modifiers: VO,
        description: "Describe the location of the insertion point (from upper-left corner of screen)",
        representation: "VO-F4-F4",
    },
    describeItemUnderMouseCursor: {
        keyCode: KeyCodes_1.KeyCodes.F5,
        modifiers: VO,
        description: "Describe the item under the mouse cursor",
        representation: "VO-F5",
    },
    describeLocationOfMouseInCoordinates: {
        keyCode: [KeyCodes_1.KeyCodes.F5, KeyCodes_1.KeyCodes.F5],
        modifiers: VO,
        description: "Describe the location of the mouse in x, y coordinates (from upper-left corner of screen)",
        representation: "VO-F5-F5",
    },
    describeLocationOfMouse: {
        keyCode: [KeyCodes_1.KeyCodes.F5, KeyCodes_1.KeyCodes.F5, KeyCodes_1.KeyCodes.F5],
        modifiers: VO,
        description: "Describe the location of the mouse (from upper-left corner of window)",
        representation: "VO-F5-F5-F5",
    },
    describeSelectedItem: {
        keyCode: KeyCodes_1.KeyCodes.F6,
        modifiers: VO,
        description: "Describe the selected item",
        representation: "VO-F6",
    },
    readCurrentSelection: {
        keyCode: KeyCodes_1.KeyCodes.a,
        modifiers: VO,
        description: "Read everything in the VoiceOver cursor",
        representation: "VO-A",
    },
    readActiveWindow: {
        keyCode: KeyCodes_1.KeyCodes.w,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read everything visible in the window or the Dock, or on your desktop, depending on your location",
        representation: "VO-Shift-W",
    },
    repeatLastSpokenPhrase: {
        keyCode: KeyCodes_1.KeyCodes.z,
        modifiers: VO,
        description: "Repeat the last spoken phrase",
        representation: "VO-Z",
    },
    copyLastSpokenPhraseToClipboard: {
        keyCode: KeyCodes_1.KeyCodes.c,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: 'Copy the last spoken phrase to the Clipboard (also called the "Pasteboard")',
        representation: "VO-Shift-C",
    },
    saveLastSpokenPhraseToDesktop: {
        keyCode: KeyCodes_1.KeyCodes.z,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Save the last spoken phrase and the crash log to a file on the desktop for troubleshooting",
        representation: "VO-Shift-Z",
    },
    // Search Commands
    find: {
        keyCode: KeyCodes_1.KeyCodes.f,
        modifiers: VO,
        description: "Find",
        representation: "VO-F",
    },
    findNextSearchedText: {
        keyCode: KeyCodes_1.KeyCodes.g,
        modifiers: VO,
        description: "Find the next searched text",
        representation: "VO-G",
    },
    findPreviousSearchedText: {
        keyCode: KeyCodes_1.KeyCodes.g,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Find the previous searched text",
        representation: "VO-Shift-G",
    },
    findNextList: {
        keyCode: KeyCodes_1.KeyCodes.x,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next list",
        representation: "VO-Command-X",
    },
    findPreviousList: {
        keyCode: KeyCodes_1.KeyCodes.x,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous list",
        representation: "VO-Command-Shift-X",
    },
    findNextBoldText: {
        keyCode: KeyCodes_1.KeyCodes.b,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next bold text",
        representation: "VO-Command-B",
    },
    findPreviousBoldText: {
        keyCode: KeyCodes_1.KeyCodes.b,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous bold text",
        representation: "VO-Command-Shift-B",
    },
    findNextStyleChange: {
        keyCode: KeyCodes_1.KeyCodes.c,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next style change",
        representation: "VO-Command-C",
    },
    findPreviousStyleChange: {
        keyCode: KeyCodes_1.KeyCodes.c,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous style change",
        representation: "VO-Command-Shift-C",
    },
    findNextItalicText: {
        keyCode: KeyCodes_1.KeyCodes.i,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next italic text",
        representation: "VO-Command-I",
    },
    findPreviousItalicText: {
        keyCode: KeyCodes_1.KeyCodes.i,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous italic text",
        representation: "VO-Command-Shift-I",
    },
    findNextColorChange: {
        keyCode: KeyCodes_1.KeyCodes.k,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next color change",
        representation: "VO-Command-K",
    },
    findPreviousColorChange: {
        keyCode: KeyCodes_1.KeyCodes.k,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous color change",
        representation: "VO-Command-Shift-K",
    },
    findNextFontChange: {
        keyCode: KeyCodes_1.KeyCodes.o,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next font change",
        representation: "VO-Command-O",
    },
    findPreviousFontChange: {
        keyCode: KeyCodes_1.KeyCodes.o,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous font change",
        representation: "VO-Command-Shift-O",
    },
    findNextTable: {
        keyCode: KeyCodes_1.KeyCodes.t,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next table",
        representation: "VO-Command-T",
    },
    findPreviousTable: {
        keyCode: KeyCodes_1.KeyCodes.t,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous table",
        representation: "VO-Command-Shift-T",
    },
    findNextUnderlinedText: {
        keyCode: KeyCodes_1.KeyCodes.u,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next underlined text",
        representation: "VO-Command-U",
    },
    findPreviousUnderlinedText: {
        keyCode: KeyCodes_1.KeyCodes.u,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous underlined text",
        representation: "VO-Command-Shift-U",
    },
    findNextControl: {
        keyCode: KeyCodes_1.KeyCodes.j,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next control",
        representation: "VO-Command-J",
    },
    findPreviousControl: {
        keyCode: KeyCodes_1.KeyCodes.j,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous control",
        representation: "VO-Command-Shift-J",
    },
    findNextDifferentItem: {
        keyCode: KeyCodes_1.KeyCodes.d,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next different item",
        representation: "VO-Command-D",
    },
    findPreviousDifferentItem: {
        keyCode: KeyCodes_1.KeyCodes.d,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous different item",
        representation: "VO-Command-Shift-D",
    },
    findNextItemWithSameTypeAsCurrentItem: {
        keyCode: KeyCodes_1.KeyCodes.s,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next item that's the same type as the current item",
        representation: "VO-Command-S",
    },
    findPreviousItemWithSameTypeAsCurrentItem: {
        keyCode: KeyCodes_1.KeyCodes.s,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous item that's the same type as the current item",
        representation: "VO-Command-Shift-S",
    },
    findNextGraphic: {
        keyCode: KeyCodes_1.KeyCodes.g,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next graphic",
        representation: "VO-Command-G",
    },
    findPreviousGraphic: {
        keyCode: KeyCodes_1.KeyCodes.g,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous graphic",
        representation: "VO-Command-Shift-G",
    },
    findNextHeading: {
        keyCode: KeyCodes_1.KeyCodes.h,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next heading",
        representation: "VO-Command-H",
    },
    findPreviousHeading: {
        keyCode: KeyCodes_1.KeyCodes.h,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous heading",
        representation: "VO-Command-Shift-H",
    },
    findNextLink: {
        keyCode: KeyCodes_1.KeyCodes.l,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next link",
        representation: "VO-Command-L",
    },
    findPreviousLink: {
        keyCode: KeyCodes_1.KeyCodes.l,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous link",
        representation: "VO-Command-Shift-L",
    },
    findNextHeadingOfSameLevel: {
        keyCode: KeyCodes_1.KeyCodes.m,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next heading of the same level",
        representation: "VO-Command-M",
    },
    findPreviousHeadingOfSameLevel: {
        keyCode: KeyCodes_1.KeyCodes.m,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous heading of the same level",
        representation: "VO-Command-Shift-M",
    },
    findNextPlainText: {
        keyCode: KeyCodes_1.KeyCodes.p,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next plain text",
        representation: "VO-Command-P",
    },
    findPreviousPlainText: {
        keyCode: KeyCodes_1.KeyCodes.p,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous plain text",
        representation: "VO-Command-Shift-P",
    },
    findNextVisitedLink: {
        keyCode: KeyCodes_1.KeyCodes.v,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next visited link",
        representation: "VO-Command-V",
    },
    findPreviousVisitedLink: {
        keyCode: KeyCodes_1.KeyCodes.v,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous visited link",
        representation: "VO-Command-Shift-V",
    },
    findNextMisspelledWord: {
        keyCode: KeyCodes_1.KeyCodes.e,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Find the next misspelled word",
        representation: "VO-Command-E",
    },
    findPreviousMisspelledWord: {
        keyCode: KeyCodes_1.KeyCodes.e,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Find the previous misspelled word",
        representation: "VO-Command-Shift-E",
    },
    // Text Commands
    readAllText: {
        keyCode: KeyCodes_1.KeyCodes.a,
        modifiers: VO,
        description: "Read all text from the VoiceOver cursor to the end of the text",
        representation: "VO-A",
    },
    selectAllText: {
        keyCode: KeyCodes_1.KeyCodes.a,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Select all text in the VoiceOver cursor",
        representation: "VO-Shift-A",
    },
    toggleTextSelection: {
        keyCode: KeyCodes_1.KeyCodes.Return,
        modifiers: VO,
        description: "Start and stop text selection in a text field (text selection tracking must be on)",
        representation: "VO-Return",
    },
    speakTextAttributes: {
        keyCode: KeyCodes_1.KeyCodes.t,
        modifiers: VO,
        description: "Speak text attributes",
        representation: "VO-T",
    },
    readParagraph: {
        keyCode: KeyCodes_1.KeyCodes.p,
        modifiers: VO,
        description: "Read paragraph in VoiceOver cursor",
        representation: "VO-P",
    },
    readNextParagraph: {
        keyCode: KeyCodes_1.KeyCodes.PageDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read next paragraph",
        representation: "VO-Shift-Page Down",
    },
    readPreviousParagraph: {
        keyCode: KeyCodes_1.KeyCodes.PageUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read previous paragraph",
        representation: "VO-Shift-Page Up",
    },
    readSentence: {
        keyCode: KeyCodes_1.KeyCodes.s,
        modifiers: VO,
        description: "Read sentence in VoiceOver cursor",
        representation: "VO-S",
    },
    readNextSentence: {
        keyCode: KeyCodes_1.KeyCodes.PageDown,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Read next sentence",
        representation: "VO-Command-Page Down",
    },
    readPreviousSentence: {
        keyCode: KeyCodes_1.KeyCodes.PageUp,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Read previous sentence",
        representation: "VO-Command-Page Up",
    },
    readLine: {
        keyCode: KeyCodes_1.KeyCodes.l,
        modifiers: VO,
        description: "Read line in VoiceOver cursor",
        representation: "VO-L",
    },
    readNextLine: {
        keyCode: KeyCodes_1.KeyCodes.ArrowDown,
        modifiers: VO,
        description: "Read next line",
        representation: "VO-Down Arrow",
    },
    readPreviousLine: {
        keyCode: KeyCodes_1.KeyCodes.ArrowUp,
        modifiers: VO,
        description: "Read previous line",
        representation: "VO-Up Arrow",
    },
    readWord: {
        keyCode: KeyCodes_1.KeyCodes.w,
        modifiers: VO,
        description: "Read word in VoiceOver cursor",
        representation: "VO-W",
    },
    readWordSpelled: {
        keyCode: [KeyCodes_1.KeyCodes.w, KeyCodes_1.KeyCodes.w],
        modifiers: VO,
        description: "Read word spelled in VoiceOver cursor",
        representation: "VO-W-W",
    },
    readWordPhonetically: {
        keyCode: [KeyCodes_1.KeyCodes.w, KeyCodes_1.KeyCodes.w, KeyCodes_1.KeyCodes.w],
        modifiers: VO,
        description: "Read word spelled phonetically in VoiceOver cursor",
        representation: "VO-W-W-W",
    },
    readNextWord: {
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: VO,
        description: "Read next word",
        representation: "VO-Right Arrow",
    },
    readPreviousWord: {
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: VO,
        description: "Read previous word",
        representation: "VO-Left Arrow",
    },
    readCharacter: {
        keyCode: KeyCodes_1.KeyCodes.c,
        modifiers: VO,
        description: "Read character in VoiceOver cursor",
        representation: "VO-W",
    },
    readCharacterPhonetically: {
        keyCode: [KeyCodes_1.KeyCodes.c, KeyCodes_1.KeyCodes.c],
        modifiers: VO,
        description: "Read character phonetically in VoiceOver cursor",
        representation: "VO-C-C",
    },
    readNextCharacter: {
        keyCode: KeyCodes_1.KeyCodes.ArrowRight,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read next character",
        representation: "VO-Shift-Right Arrow",
    },
    readPreviousCharacter: {
        keyCode: KeyCodes_1.KeyCodes.ArrowLeft,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read previous character",
        representation: "VO-Shift-Left Arrow",
    },
    moveToFirstVisibleWord: {
        keyCode: KeyCodes_1.KeyCodes.Home,
        modifiers: VO,
        description: "Move to first visible word",
        representation: "VO-Home",
    },
    moveToLastVisibleWord: {
        keyCode: KeyCodes_1.KeyCodes.End,
        modifiers: VO,
        description: "Move to last visible word",
        representation: "VO-End",
    },
    moveToBeginningOfText: {
        keyCode: KeyCodes_1.KeyCodes.Home,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to beginning of text, scrolling if necessary",
        representation: "VO-Shift-Home",
    },
    moveToEndOfText: {
        keyCode: KeyCodes_1.KeyCodes.End,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Move to end of text, scrolling if necessary",
        representation: "VO-Shift-End",
    },
    readCurrentWordAndCharacter: {
        keyCode: KeyCodes_1.KeyCodes.F3,
        modifiers: VO,
        description: "Reads the current word and character in the VoiceOver cursor",
        representation: "VO-F3",
    },
    readNumberOfLines: {
        keyCode: [KeyCodes_1.KeyCodes.F3, KeyCodes_1.KeyCodes.F3],
        modifiers: VO,
        description: "Reads the total number of lines and the number of visible lines in a document",
        representation: "VO-F3-F3",
    },
    // Web Commands
    moveToNextColumn: {
        keyCode: KeyCodes_1.KeyCodes.y,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the next column",
        representation: "VO-Command-Y",
    },
    moveToPreviousColumn: {
        keyCode: KeyCodes_1.KeyCodes.y,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Move to the previous column",
        representation: "VO-Command-Shift-Y",
    },
    moveToNextFrame: {
        keyCode: KeyCodes_1.KeyCodes.f,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the next frame",
        representation: "VO-Command-F",
    },
    moveToPreviousFrame: {
        keyCode: KeyCodes_1.KeyCodes.f,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Move to the previous frame",
        representation: "VO-Command-Shift-F",
    },
    moveToNextAutoWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.n,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the next auto web spot",
        representation: "VO-Command-N",
    },
    moveToPreviousAutoWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.n,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Move to the previous auto web spot",
        representation: "VO-Command-Shift-N",
    },
    moveToNextWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.LeftSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the next web spot",
        representation: "VO-Command-]",
    },
    moveToPreviousWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.RightSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Command],
        description: "Move to the previous web spot",
        representation: "VO-Command-[",
    },
    openWebItemRotor: {
        keyCode: KeyCodes_1.KeyCodes.u,
        modifiers: VO,
        description: "Open the Web Item rotor",
        representation: "VO-U",
    },
    readFromBeginningToCurrent: {
        keyCode: KeyCodes_1.KeyCodes.b,
        modifiers: VO,
        description: "Read from the beginning of a webpage to the current location",
        representation: "VO-B",
    },
    readLinkAddress: {
        keyCode: KeyCodes_1.KeyCodes.u,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read a link address (URL)",
        representation: "VO-Shift-U",
    },
    readWebpageStatistics: {
        keyCode: KeyCodes_1.KeyCodes.i,
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Read webpage statistics",
        representation: "VO-Shift-I",
    },
    removeWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.LeftSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Remove a web spot",
        representation: "VO-Command-Shift-{",
    },
    setWebSpot: {
        keyCode: KeyCodes_1.KeyCodes.RightSquareBracket,
        modifiers: [...VO, Modifiers_1.Modifiers.Command, Modifiers_1.Modifiers.Shift],
        description: "Set a web spot",
        representation: "VO-Command-Shift-}",
    },
    setSweetSpot: {
        keyCode: [KeyCodes_1.KeyCodes.RightSquareBracket, KeyCodes_1.KeyCodes.RightSquareBracket],
        modifiers: [...VO, Modifiers_1.Modifiers.Shift],
        description: "Set the sweet spot",
        representation: "VO-Command-Shift-}-}",
    },
    toggleGroupingItemsWithinTable: {
        keyCode: KeyCodes_1.KeyCodes.Equals,
        modifiers: VO,
        description: "Turn the grouping of items within a table on or off",
        representation: "VO-=",
    },
};
